nmdc_automation/models/nmdc.py

```diff
@@ -0,0 +1,116 @@
+WorkflowExecution]:
+    """
+    Factory function to create a PlannedProcess subclass object from a record.
+    Subclasses are determined by the "type" field in the record, and can be
+    either a WorkflowExecution or DataGeneration object.
+    """
+    nmdc_materialized = get_nmdc_materialized()
+    process_types = {
+        "nmdc:MagsAnalysis": MagsAnalysis,
+        "nmdc:MetagenomeAnnotation": MetagenomeAnnotation,
+        "nmdc:MetagenomeAssembly": MetagenomeAssembly,
+        "nmdc:MetatranscriptomeAnnotation": MetatranscriptomeAnnotation,
+        "nmdc:MetatranscriptomeAssembly": MetatranscriptomeAssembly,
+        "nmdc:MetatranscriptomeExpressionAnalysis": MetatranscriptomeExpressionAnalysis,
+        "nmdc:NucleotideSequencing": NucleotideSequencing,
+        "nmdc:ReadBasedTaxonomyAnalysis": ReadBasedTaxonomyAnalysis,
+        "nmdc:ReadQcAnalysis": ReadQcAnalysis,
+    }
+    record = _normalize_record(record)
+    target_class = record["type"].split(":")[1]
+    if validate:
+        validation_report = linkml.validator.validate(record, nmdc_materialized, target_class)
+        if validation_report.results:
+            raise ValueError(f"Validation error: {validation_report.results[0].message}")
+
+
+
+
+    try:
+        cls = process_types[record["type"]]
+    except KeyError:
+        raise ValueError(f"Invalid workflow execution type: {record['type']}")
+    wfe = cls(**record)
+    return wfe
+
+
+def _normalize_record(record: Dict[str, Any]) -> Dict[str, Any]:
+    """ Normalize the record by removing the _id field and converting the type field to a string """
+    record.pop("_id", None)
+    # for backwards compatibility strip Activity from the end of the type
+    record["type"] = record["type"].replace("Activity", "")
+    normalized_record = _strip_empty_values(record)
+
+
+    # type-specific normalization
+    if normalized_record["type"] == "nmdc:MagsAnalysis":
+        normalized_record = _normalize_mags_record(normalized_record)
+
+    return normalized_record
+
+
+def _normalize_mags_record(record: Dict[str, Any]) -> Dict[str, Any]:
+    """ Normalize the record for a MagsAnalysis object """
+    for i, mag in enumerate(record.get("mags_list", [])):
+        if not mag.get("type"):
+            # Update the original dictionary in the list
+            record["mags_list"][i]["type"] = "nmdc:MagBin"
+        # for backwards compatibility normalize num_tRNA to num_t_rna
+        if "num_tRNA" in mag:
+            record["mags_list"][i]["num_t_rna"] = mag.pop("num_tRNA")
+        # add type to eukaryotic_evaluation if it exists
+        if "eukaryotic_evaluation" in mag:
+            record["mags_list"][i]["eukaryotic_evaluation"]["type"] = "nmdc:EukEval"
+    return record
+
+
+def _strip_empty_values(d: Dict[str, Any]) -> Dict[str, Any]:
+    """ Strip empty values from a record """
+    empty_values = [None, "", [], "null",]
```

Comment:
think about if we want to keep null. We need some way to distinguish between not assigned and they key doesn't exist because of other issues. Use case is taxonomic assignment. is the species key not there b/c of bug or because it couldn't be assigned to that granularity.  cc @turbomam @cmungall

Reply:
https://github.com/microbiomedata/nmdc_automation/pull/274/commits/8da0ceb007b27eda679039ebcca2291458b54147